services/core/java/com/android/server/wm/WindowManagerHooks.java

```diff
@@ -0,0 +1,71 @@
+package com.android.server.wm;
+
+import android.app.ActivityThread;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.content.pm.PackageManagerInternal;
+import android.ext.BrowserUtils;
+import android.os.Binder;
+import android.os.Build;
+import android.os.SystemProperties;
+import android.os.UserHandle;
+
+import com.android.internal.util.ArrayUtils;
+
+import com.android.server.LocalServices;
+import com.android.server.pm.permission.PermissionManagerServiceInternal;
+import com.android.server.pm.pkg.AndroidPackage;
+import com.android.server.pm.pkg.PackageStateInternal;
+
+class WindowManagerHooks {
+
+    static boolean canAccessLaunchedFromPackagePermission() {
+        final int callingUid = Binder.getCallingUid();
+        var pmi = LocalServices.getService(PackageManagerInternal.class);
+        AndroidPackage callingPkg = pmi.getPackage(callingUid);
+        if (callingPkg == null) {
+            return false;
+        }
+
+        String callingPkgName = callingPkg.getPackageName();
+        if (canAccessForDebuggingPurposes(callingPkgName)) {
+            return true;
+        }
+
+        Context ctx = ActivityThread.currentActivityThread().getSystemContext();
+        if (!BrowserUtils.isSystemBrowser(ctx, callingPkgName)) {
+            return false;
+        }
+
+        PackageStateInternal callingPsi = pmi.getPackageStateInternal(callingPkg.getPackageName());
+        if (callingPsi == null) {
+            return false;
+        }
+
+        return callingPsi.isSystem();
+    }
+
+    static boolean canAccessForDebuggingPurposes(String packageName) {
+        if (!Build.isDebuggable()) {
+            return false;
+        }
+
+        String testPkgs = SystemProperties.get("persist.launchedFromPackagePermission_test_pkgs");
+        return ArrayUtils.contains(testPkgs.split(","), packageName);
+    }
+
+    static boolean doesLaunchedFromPackageHavePermission(ActivityRecord r, String permission) {
+        String packageName = r.launchedFromPackage;
+        final int userId = UserHandle.getUserId(r.launchedFromUid);
+        var permService = LocalServices.getService(PermissionManagerServiceInternal.class);
+
+        // Do not take into account of package visibility here.
+        long token = Binder.clearCallingIdentity();
+        try {
+            return permService.checkPermission(packageName, permission, userId) == PackageManager.PERMISSION_GRANTED;
```

Comment:
It's better to return `checkPermission()` result directly, there might be more return values in the future.